### PR TITLE
[TVMC] Fix PyTorch support

### PR DESCRIPTION
**Pull request by ekalda**
A PyTorch model could not be compiled through tvmc because the shape
of the input tensor could not be deduced from the model after it has been
saved. We've added an --input-shape parameter to tvmc compile and
tvmc tune that allows the inputs to be specified for PyTorch models.

### CODE REVIEW

**Comment by ekalda** [comment]
cc @leandron @u99127

File: python/tvm/driver/tvmc/common.py [hunk 1]
```diff
@@ -136,3 +138,40 @@ def tracker_host_port_from_cli(rpc_tracker_str):
         logger.info("RPC tracker port: %s", rpc_port)
 
     return rpc_hostname, rpc_port
+
+
+def parse_input_shapes(xs):
+    """Turn the string from --input-shape into a list.
+
```
Comment (leandron) [inline python/tvm/driver/tvmc/common.py:145]: It would be good to have an example here, that describes the input format and expected output format, similar to what you have on `test_parse_input_shapes__turn_into_list`.

File: python/tvm/driver/tvmc/compiler.py [hunk 2]
```diff
@@ -59,6 +59,12 @@ def add_compile_parser(subparsers):
         default="",
         help="comma separarated list of formats to export, e.g. 'asm,ll,relay' ",
     )
+    parser.add_argument(
+        "--input-shape",
+        type=common.parse_input_shapes,
+        metavar="INPUT_SHAPE,[INPUT_SHAPE]...",
+        help="for PyTorch, e.g. '(1,3,224,224)'",
```
Comment (leandron) [inline python/tvm/driver/tvmc/compiler.py:66]: Maybe clarify that it is in fact mandatory for PyTorch.

**Review by leandron — COMMENTED** [review]
Thanks for this fix.

cc @masahi @comaniac to have a look.

File: python/tvm/driver/tvmc/compiler.py (hunk 2 above)
Comment (comaniac) [reply python/tvm/driver/tvmc/compiler.py:66]: Agree. It's confusing to see such a general option only for PyTorch. I would suggest the following changes:
1. Make `--input-shape` as a general option for all frontends. If present, we skip the input shape inference.
2. `--input-shape` is optional by default. However, if users want to process a PyTorch model but don't specify `--input-shape`, we throw out an error in the PyTorch frontend.

File: python/tvm/driver/tvmc/common.py [hunk 3]
```diff
@@ -136,3 +138,40 @@ def tracker_host_port_from_cli(rpc_tracker_str):
...
+    shapes = []
+    # Split up string into comma seperated sections ignoring commas in ()s
+    match = re.findall(r"(\(.*?\)|.+?),?", xs)
+    if match:
+        for inp in match:
+            # Test for and remove brackets
+            shape = re.match(r"\((.*)\)", inp)
+            if shape and shape.lastindex == 1:
+                # Remove white space and extract numbers
+                strshape = shape[1].replace(" ", "").split(",")
```
Comment (comaniac) [inline python/tvm/driver/tvmc/common.py:166]: It would be safer and easier to remove all spaces in `xs` in the beginning of this function.

File: python/tvm/driver/tvmc/common.py [hunk 4]
```diff
@@ -136,3 +138,40 @@ def tracker_host_port_from_cli(rpc_tracker_str):
...
+                try:
+                    shapes.append([int(i) for i in strshape])
+                except ValueError:
+                    raise argparse.ArgumentTypeError(f"expected numbers in shape '{shape[1]}'")
```
Comment (comaniac) [inline python/tvm/driver/tvmc/common.py:170]: Consider the following two input shapes:
- `(8)`: `shapes=[8]`
- `(8,)`: Value error because `strshape` would be `[8, ""]`.

Accordingly, I guess your intention is `(8)` instead of `(8,)`. However, this is inconsistent with the Python syntax so it might confuse people. I have two proposals to deal with this:
1. Use list syntax instead of tuple, so that the semantic is clear, and we can simply use JSON loader to deal with all variants (e.g., spaces):
    ```python
    xs = "[1,3,224,224], [32]"
    shapes = json.loads(xs) # [[1,3,224,224],[32]]
    ```
2. Follow Python syntax to only accept `(8,)` and throw an error for `(8)`, which is treated as an integer instead of a tuple because buckets will be simplified in Python. In this case, I would suggest using `eval` to deal with all variants.
    ```python
    xs = "(1,3,224,224), (32,)"
    shapes = eval(xs, {}, {}) # Remember to disable all local and global symbols to isolate this expression.
    # shapes=[(1,3,224,224),(32,)]
    ```

Either way is fine for me, and please update the help message and make sure you have a unit test to cover corner cases.

File: python/tvm/driver/tvmc/frontends.py [hunk 5]
```diff
@@ -99,10 +101,13 @@ def name():
...
         return ["h5"]
 
-    def load(self, path):
+    def load(self, path, input_shape):
         # pylint: disable=C0103
         tf, keras = import_keras()
 
+        if input_shape:
+            raise TVMCException("--input-shape is not supported for {}".format(self.name()))
+
```
Comment (comaniac) [inline python/tvm/driver/tvmc/frontends.py:110]: This is definitely too ad hoc

File: python/tvm/driver/tvmc/frontends.py [hunk 6]
```diff
@@ -285,17 +299,17 @@ def suffixes():
...
         return ["pth", "zip"]
 
-    def load(self, path):
+    def load(self, path, input_shape):
         # pylint: disable=C0415
         import torch
 
-        traced_model = torch.jit.load(path)
-
-        inputs = list(traced_model.graph.inputs())[1:]
```
Comment (comaniac) [inline python/tvm/driver/tvmc/frontends.py:294]: Is this approach not working at all? If it works for some cases, we should still use it first when `--input-shape` is missing.

File: python/tvm/driver/tvmc/frontends.py [hunk 7]
```diff
@@ -389,6 +403,8 @@ def load_model(path, model_format=None):
     model_format : str, optional
         The underlying framework used to create the model.
         If not specified, this will be inferred from the file type.
+    input shape : list, optional
+        The shape of input tensor for PyTorch models
```
Comment (comaniac) [inline python/tvm/driver/tvmc/frontends.py:407]: ditto. make it general instead of only for PyTorch.

**Review by comaniac — CHANGES_REQUESTED** [review]

File: python/tvm/driver/tvmc/frontends.py (hunk 6 above)
Comment (ekalda) [reply python/tvm/driver/tvmc/frontends.py:294]: I looked into this and I didn't find a way to extract inputs from the model after it has been saved and loaded. I asked on the PyTorch forum as well (https://discuss.pytorch.org/t/input-size-disappears-between-torch-jit-save-and-torch-jit-load/108955) and since I received a grand total of zero responses, I suspect it is a deliberate design decision. If there was a way, it would be good to keep it, of course, but in that form it doesn't work any more.

**Comment by comaniac** [comment]
Include the functionalities in #7366.
